File: backend/setup_db.py
import os
import mysql.connector
from dotenv import load_dotenv
import bcrypt

load_dotenv()

def initialize_database():
    print("Connecting to MySQL...")
    try:
        # Connect without database first to create it
        conn = mysql.connector.connect(
            host=os.environ.get("DB_HOST", "127.0.0.1"),
            port=os.environ.get("DB_PORT", "3306"),
            user=os.environ.get("DB_USER", "job_user"),
            password=os.environ.get("DB_PASSWORD", "job_password")
        )
        cursor = conn.cursor()
        
        # Read and execute schema.sql
        print("Executing schema.sql...")
        schema_path = os.path.join(os.path.dirname(__file__), 'schema.sql')
        with open(schema_path, 'r') as f:
            schema_sql = f.read()
            
        # Execute statements individually
        sql_commands = [cmd.strip() for cmd in schema_sql.split(';') if cmd.strip()]
        for command in sql_commands:
            cursor.execute(command)
            
        print("Database schema created successfully.")
        
        # Seed Data
        cursor.execute("USE job_portal;")
        
        # Check if users already exist
        cursor.execute("SELECT COUNT(*) FROM users;")
        if cursor.fetchone()[0] == 0:
            print("Inserting seed data...")
            hashed_pw = bcrypt.hashpw('password123'.encode('utf-8'), bcrypt.gensalt()).decode('utf-8')
            
            # Insert Employer
            cursor.execute("""
                INSERT INTO users (name, email, password_hash, role) 
                VALUES ('Acme Corp', 'employer@acme.com', %s, 'employer')
            """, (hashed_pw,))
            employer_id = cursor.lastrowid
            
            # Insert Job Seeker
            cursor.execute("""
                INSERT INTO users (name, email, password_hash, role) 
                VALUES ('John Doe', 'john@example.com', %s, 'seeker')
            """, (hashed_pw,))
            seeker_id = cursor.lastrowid
            
            # Insert Job
            cursor.execute("""
                INSERT INTO jobs (employer_id, title, company, description, requirements, salary, location, type) 
                VALUES (%s, 'Software Engineer', 'Acme Corp', 'Develop awesome stuff', 'Python, MySQL', '$100k - $120k', 'Remote', 'Full-time')
            """, (employer_id,))
            job_id = cursor.lastrowid
            
            # Insert Application
            cursor.execute("""
                INSERT INTO applications (job_id, user_id, resume_link, status) 
                VALUES (%s, %s, 'http://example.com/resume.pdf', 'pending')
            """, (job_id, seeker_id))
            
            conn.commit()
            print("Seed data inserted successfully.")
            print(f"Employer: employer@acme.com / password123")
            print(f"Seeker: john@example.com / password123")
        else:
            print("Database already contains data. Skipping seed.")
            
    except mysql.connector.Error as err:
        print(f"Error: {err}")
    finally:
        if 'cursor' in locals():
            cursor.close()
        if 'conn' in locals():
            conn.close()

if __name__ == "__main__":
    initialize_database()

File: backend/setupDb.js
const mysql = require('mysql2/promise');
const fs = require('fs');
const path = require('path');
const bcrypt = require('bcryptjs');
const dotenv = require('dotenv');

dotenv.config();

async function initializeDatabase() {
  console.log('Connecting to MySQL...');
  let connection;
  try {
    // Connect without DB first
    connection = await mysql.createConnection({
      host: process.env.DB_HOST || '127.0.0.1',
      port: process.env.DB_PORT || 3306,
      user: process.env.DB_USER || 'root',
      password: process.env.DB_PASSWORD || '',
      multipleStatements: true
    });

    console.log('Executing schema.sql...');
    const schemaPath = path.join(__dirname, 'schema.sql');
    const schemaSql = fs.readFileSync(schemaPath, 'utf8');

    // Split and execute statements one by one for safety
    const sqlCommands = schemaSql.split(';').map(cmd => cmd.trim()).filter(cmd => cmd.length > 0);
    for (let command of sqlCommands) {
      await connection.query(command);
    }
    
    console.log('Database schema created successfully.');

    // Switch to the database for seeding
    await connection.query('USE job_portal');

    // Check if users exist
    const [rows] = await connection.query('SELECT COUNT(*) as count FROM users');
    if (rows[0].count === 0) {
      console.log('Inserting seed data...');
      
      const salt = await bcrypt.genSalt(10);
      const hashedPassword = await bcrypt.hash('password123', salt);

      // Insert Employer
      const [empResult] = await connection.query(
        "INSERT INTO users (name, email, password_hash, role) VALUES ('Acme Corp', 'employer@acme.com', ?, 'employer')",
        [hashedPassword]
      );
      const employerId = empResult.insertId;

      // Insert Seeker
      const [seekerResult] = await connection.query(
        "INSERT INTO users (name, email, password_hash, role) VALUES ('John Doe', 'john@example.com', ?, 'seeker')",
        [hashedPassword]
      );
      const seekerId = seekerResult.insertId;

      // Insert Job
      const [jobResult] = await connection.query(
        "INSERT INTO jobs (employer_id, title, company, description, requirements, salary, location, type) VALUES (?, 'Software Engineer', 'Acme Corp', 'Develop awesome stuff', 'Node.js, MySQL', '$100k - $120k', 'Remote', 'Full-time')",
        [employerId]
      );
      const jobId = jobResult.insertId;

      // Insert Application
      await connection.query(
        "INSERT INTO applications (job_id, user_id, resume_link, status) VALUES (?, ?, 'http://example.com/resume.pdf', 'pending')",
        [jobId, seekerId]
      );

      console.log('Seed data inserted successfully.');
      console.log(`Employer: employer@acme.com / password123`);
      console.log(`Seeker: john@example.com / password123`);
    } else {
      console.log('Database already contains data. Skipping seed.');
    }

  } catch (err) {
    console.error('Error:', err.message);
  } finally {
    if (connection) {
      await connection.end();
    }
    process.exit();
  }
}

initializeDatabase();
